arxiv/taxonomy/category.py

```diff
@@ -1,73 +1,90 @@
-            catname = self.lookup[self]['name']
-            return f'{catname} ({self})'
-
-        parts = self.split('.', 2)
-        if len(parts) == 2:    # Has subject.
-            (archive, _) = parts
-            if archive in ARCHIVES:
-                archive_name = ARCHIVES[archive]['name']
-                return f'{archive_name} ({archive})'
-        return self
-
-    def unalias(self) -> 'Category':
-        """Follow any EQUIV or SUBSUMED to get the current category."""
-        if self in CATEGORY_ALIASES:
-            return Category(CATEGORY_ALIASES[self])
-        if self in ARCHIVES_SUBSUMED:
-            return Category(ARCHIVES_SUBSUMED[self])
-        return self
+        from .definitions import CATEGORIES,ARCHIVES, GROUPS
+        if not canonical:
+            return f'{self.full_name} ({self.id})'
+        else:
+            name=self.canonical 
+            if name==self.id:
+                return f'{self.full_name} ({self.id})'
+            elif name in CATEGORIES.keys():
+                item=CATEGORIES[name]
+                return f'{item.full_name} ({item.id})'
+            elif name in ARCHIVES.keys():#will never happen with taxonomy as of 2024
+                item=ARCHIVES[name]
+                return f'{item.full_name} ({item.id})'
+            elif name in GROUPS.keys(): #will never happen with taxonomy as of 2024
+                item=GROUPS[name]
+                return f'{item.full_name} ({item.id})'
+
+class Group(BaseTaxonomy):
+    """Represents an arXiv group--the highest (most general) taxonomy level."""
+    start_year: int
+    default_archive: Optional[str]
+    is_test: Optional[bool]
 
+    def get_archives(self) -> List['Archive'] :
+        """creates a list of all archives withing the group"""
+        from .definitions import ARCHIVES
+        return [archive for archive in ARCHIVES.values() if archive.in_group == self.id]
 
-class Archive(Category):
+class Archive(BaseTaxonomy):
     """Represents an arXiv archive--the middle level of the taxonomy."""
 
-    lookup = ARCHIVES
-
+    in_group: str
+    start_date: date
+    end_date: Optional[date]
+
+    def get_group(self) -> Group:
+        """Returns parent archive."""
+        from .definitions import GROUPS
+        return GROUPS[self.in_group]
+    
+    def get_categories(self) -> List['Category'] :
+        """creates a list of all categories withing the group"""
+        from .definitions import CATEGORIES,ARCHIVES, GROUPS
+        return [category for category in CATEGORIES.values() if category.in_archive == self.id]
+
+class Category(BaseTaxonomy):
+    """Represents an arXiv category."""
 
-class Group(Category):
-    """Represents an arXiv group--the highest (most general) taxonomy level."""
+    in_archive: str
+    is_general: bool
+    description: Optional[str]
 
-    lookup = GROUPS
+    def get_archive(self) -> Archive:
+        """Returns parent archive."""
+        from .definitions import ARCHIVES
```

Comment:
Import at the top unless there's a good reason not to

Reply:
its circular if you do it at the top :/ ARCHIVES relies on the structure in taxonomy to define everything. But I still needed access to ARCHIVES within some of the functions the class uses

Reply:
Oof yeah. Since they are codependent, it may make sense to put them in the same file. I think the taxonomy dir should have a __init__.py too?

Reply:
Alternatively, those methods can be module methods instead of class methods like `def get_archive (group: Group) -> Archive:`

Reply:
As far as putting them in the same file, there's still an issue with what order you define them in. And honestly the definitions file is huge and I wanted the category file with type definitions to be more readable/usable